### ISSUE TITLE
partial rebuilding

### ISSUE THREAD

**Issue by pelikhan** [lancaster-university/codal#33]
If I execute ``python build.py`` immediately after itself, I would have expected a much faster build time. It seems to be rebuilding all the files all over again.

**Comment by finneyj** [comment]
Yes, this seems to be a windows feature - my build suffers from this also. Something in there is forcing a clean rebuild... @jamesadevine is it possible to test a ninja-less build with just cmake?

**Comment by jamesadevine** [comment]
Nothing forces a clean rebuild. The script enters the build folder and executes `ninja`.

**Comment by finneyj** [comment]
Is there a way to not use ninja though? As a means to honing in on the cause of the problem?

**Comment by jamesadevine** [comment]
Ah I see, you can use whatever build system you want...

```cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -G \"Ninja\"```

Change Ninja to any other supported one listed here: 

https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html#command-line-build-tool-generators

There are not many options for windows, and make runs really slowly. 

I imagine this is specific to ninja as partial rebuilding is supported on Mac using make.

`ninja -d explain` will help diagnose the cause.

**Comment by finneyj** [comment]
Ta. Yes, just looking. make doesn't run slowly compared to a clean rebuild all the time . ;)

I've just been digging and it seems to come down to one header file that (I guess) is being regenerated every build, which in turn makes everything look dirty. My windows logs are full of:

```

**Comment by finneyj** [comment]
I'll have a look to see if we can do something more optimised with that header file.

**Comment by jamesadevine** [comment]
ok - quick test then... Could you just run `ninja` in the build folder, and see if that partially rebuilds?

**Comment by finneyj** [comment]
already tried that - it doesn't rebuild, just relinks the final binaries...

Anything magical about "codal_extra_definitions.h" in the build process?

**Comment by jamesadevine** [comment]
Could you replace the similar looking code at line 175, with this:
```
# create a header file from the definitions specified in JSON
if("${CODAL_DEFINITIONS}" STRGREATER "")
    set(EXTRA_INCLUDES_NEW_PATH "${PROJECT_SOURCE_DIR}/build/codal_extra_definitions_new.h")
    set(EXTRA_INCLUDES_PATH "${PROJECT_SOURCE_DIR}/build/codal_extra_definitions.h")
    file(WRITE "${EXTRA_INCLUDES_NEW_PATH}" ${CODAL_DEFINITIONS})
    configure_file(${EXTRA_INCLUDES_PATH} ${EXTRA_INCLUDES_NEW_PATH} COPYONLY)

    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -include ${EXTRA_INCLUDES_PATH}")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -include ${EXTRA_INCLUDES_PATH}")
endif()
```

And report back?

**Comment by finneyj** [comment]
yes, I was just looking slightly askew at that code. Will try now.

**Comment by jamesadevine** [comment]
It looks like make does its' diff process differently to ninja.

**Comment by finneyj** [comment]
yup - that fixes it for me.

**Comment by jamesadevine** [comment]
Great, will push.

**Comment by finneyj** [comment]
Thanks James.

**Comment by jamesadevine** [comment]
https://github.com/lancaster-university/codal/commit/281e03d5beba89775b29b7c1f64d335313f5c2b1
